MathlibTest/depRewrite.lean

```diff
@@ -0,0 +1,199 @@
+import Mathlib.Tactic.DepRewrite
+
+/-! Basic tests for `rewrite!`. -/
+
+private axiom test_sorry : ∀ {α}, α
+
+/-- Turn a term into a sort for testing. -/
+private axiom P.{u} {α : Sort u} : α → Prop
+
+/-- Non-deprecated copy of `Fin.ofNat` for testing. -/
+private def finOfNat (n : Nat) (a : Nat) : Fin (n + 1) :=
+  ⟨a % (n+1), Nat.mod_lt _ (Nat.zero_lt_succ _)⟩
+
+open Lean Elab Term in
+/-- Produce the annotation ``.mdata .. e`` for testing.
+Standard elaborators may represent projections as ordinary functions instead. -/
+elab "mdata% " e:term : term => do
+  let e ← elabTerm e none
+  return .mdata ({ : MData}.insert `abc "def") e
+
+open Lean Elab Meta Term in
+/-- Produce the projection ``.proj `Prod 0 e`` for testing.
+Standard elaborators may represent projections as ordinary functions instead. -/
+elab "fst% " e:term : term => do
+  let u ← mkFreshLevelMVar
+  let v ← mkFreshLevelMVar
+  let α ← mkFreshExprMVar (some <| .sort u)
+  let β ← mkFreshExprMVar (some <| .sort v)
+  let tp := mkApp2 (.const ``Prod [u, v]) α β
+  let e ← elabTerm e tp
+  return .proj ``Prod 0 e
+
+/-! Tests for proof-only mode. -/
+
+variable {n m : Nat} (eq : n = m)
+
+-- mdata
+example (f : (k : Nat) → 0 < k → Type) (lt : 0 < n) : P (mdata% f n lt) := by
+  rewrite! [eq]
+  guard_target =ₐ P (mdata% f m (eq ▸ lt))
+  exact test_sorry
+
+-- app (fn)
+/-- error: function expected
```

Comment:
I imagine this is from `throwFunctionExpected` in `visit`. Is it possible to make a more-informative error message?